# PROYECTO INDIVIDUAL Nº1

***Machine Learning Operations (MLOps)***

**Proceso ETL y EDA**

**1 - Importación de librerías**

In [547]:
import pandas as pd
import numpy as np
import ast
import pyarrow as pa
import pyarrow.parquet as pq

**2 - Lectura del archivo JSON y creación del Data Frame**

In [548]:
lista_juegos = []
steam_games = 'steam_games.json'
with open(steam_games, 'r') as archivo:
    for linea in archivo.readlines():                   #Metodo para leer las líneas del JSON
        lista_juegos.append(ast.literal_eval(linea))    #Función del modulo ast.  Evalúa un literal de datos básicos de Pyhton,
                                                        #como un diccionario, una lista, una cadena, un número, un conjunto o una tupla
df0 = pd.DataFrame(lista_juegos)                        #df0 data frame del proyecto (Rev. 0)

**3 – Características y calidad de datos**

3.a - Consulta de las características de Data Frame

Son 32135 registros y 16 campos.

In [549]:
df0.shape    #Consulta de las características de dataset. Tiene 32135 registros y 16 campos

(32135, 16)

In [550]:
df0.head()   #Consulta de la estructura del dataset

publisher                                             genres  \
0         Kotoshiro      [Action, Casual, Indie, Simulation, Strategy]   
1  Making Fun, Inc.               [Free to Play, Indie, RPG, Strategy]   
2      Poolians.com  [Casual, Free to Play, Indie, Simulation, Sports]   
3              彼岸领域                        [Action, Adventure, Casual]   
4               NaN                                                NaN   

                  app_name                    title  \
0      Lost Summoner Kitty      Lost Summoner Kitty   
1                Ironbound                Ironbound   
2  Real Pool 3D - Poolians  Real Pool 3D - Poolians   
3                  弹炸人2222                  弹炸人2222   
4            Log Challenge                      NaN   

                                                 url release_date  \
0  http://store.steampowered.com/app/761140/Lost_...   2018-01-04   
1  http://store.steampowered.com/app/643980/Ironb...   2018-01-04   
2  http://store.steampowered.com/app/670290/Real_...   2017-07-24   
3     http://store.steampowered.com/app/767400/2222/   2017-12-07   
4  http://store.steampowered.com/app/773570/Log_C...          NaN   

                                                tags  discount_price  \
0      [Strategy, Action, Indie, Casual, Simulation]            4.49   
1  [Free to Play, Strategy, Indie, RPG, Card Game...             NaN   
2  [Free to Play, Simulation, Sports, Casual, Ind...             NaN   
3                        [Action, Adventure, Casual]            0.83   
4                    [Action, Indie, Casual, Sports]            1.79   

                                         reviews_url  \
0  http://steamcommunity.com/app/761140/reviews/?...   
1  http://steamcommunity.com/app/643980/reviews/?...   
2  http://steamcommunity.com/app/670290/reviews/?...   
3  http://steamcommunity.com/app/767400/reviews/?...   
4  http://steamcommunity.com/app/773570/reviews/?...   

                                               specs         price  \
0                                    [Single-player]          4.99   
1  [Single-player, Multi-player, Online Multi-Pla...  Free To Play   
2  [Single-player, Multi-player, Online Multi-Pla...  Free to Play   
3                                    [Single-player]          0.99   
4  [Single-player, Full controller support, HTC V...          2.99   

   early_access      id         developer        sentiment metascore  
0         False  761140         Kotoshiro              NaN       NaN  
1         False  643980  Secret Level SRL  Mostly Positive       NaN  
2         False  670290      Poolians.com  Mostly Positive       NaN  
3         False  767400              彼岸领域              NaN       NaN  
4         False  773570               NaN              NaN       NaN

In [551]:
df0.tail()

publisher                                 genres  \
32130  Ghost_RUS Games  [Casual, Indie, Simulation, Strategy]   
32131           Sacada              [Casual, Indie, Strategy]   
32132     Laush Studio            [Indie, Racing, Simulation]   
32133         SIXNAILS                        [Casual, Indie]   
32134              NaN                                    NaN   

                       app_name                     title  \
32130            Colony On Mars            Colony On Mars   
32131  LOGistICAL: South Africa  LOGistICAL: South Africa   
32132             Russian Roads             Russian Roads   
32133       EXIT 2 - Directions       EXIT 2 - Directions   
32134               Maze Run VR                       NaN   

                                                     url release_date  \
32130  http://store.steampowered.com/app/773640/Colon...   2018-01-04   
32131  http://store.steampowered.com/app/733530/LOGis...   2018-01-04   
32132  http://store.steampowered.com/app/610660/Russi...   2018-01-04   
32133  http://store.steampowered.com/app/658870/EXIT_...   2017-09-02   
32134  http://store.steampowered.com/app/681550/Maze_...          NaN   

                                                    tags  discount_price  \
32130              [Strategy, Indie, Casual, Simulation]            1.49   
32131                          [Strategy, Indie, Casual]            4.24   
32132                        [Indie, Simulation, Racing]            1.39   
32133  [Indie, Casual, Puzzle, Singleplayer, Atmosphe...             NaN   
32134  [Early Access, Adventure, Indie, Action, Simul...             NaN   

                                             reviews_url  \
32130  http://steamcommunity.com/app/773640/reviews/?...   
32131  http://steamcommunity.com/app/733530/reviews/?...   
32132  http://steamcommunity.com/app/610660/reviews/?...   
32133  http://steamcommunity.com/app/658870/reviews/?...   
32134  http://steamcommunity.com/app/681550/reviews/?...   

                                                   specs price  early_access  \
32130                [Single-player, Steam Achievements]  1.99         False   
32131  [Single-player, Steam Achievements, Steam Clou...  4.99         False   
32132  [Single-player, Steam Achievements, Steam Trad...  1.99         False   
32133   [Single-player, Steam Achievements, Steam Cloud]  4.99         False   
32134  [Single-player, Stats, Steam Leaderboards, HTC...  4.99          True   

           id                 developer       sentiment metascore  
32130  773640        Nikita "Ghost_RUS"             NaN       NaN  
32131  733530                    Sacada             NaN       NaN  
32132  610660  Laush Dmitriy Sergeevich             NaN       NaN  
32133  658870             xropi,stev3ns  1 user reviews       NaN  
32134  681550                       NaN        Positive       NaN

3.b – Consulta de las cantidades de registros por campos y tipos de datos

In [552]:
df0.info() #Consulta de las cantidades de registros por campos y tipos de datos

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32135 entries, 0 to 32134
Data columns (total 16 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   publisher       24083 non-null  object 
 1   genres          28852 non-null  object 
 2   app_name        32133 non-null  object 
 3   title           30085 non-null  object 
 4   url             32135 non-null  object 
 5   release_date    30068 non-null  object 
 6   tags            31972 non-null  object 
 7   discount_price  225 non-null    float64
 8   reviews_url     32133 non-null  object 
 9   specs           31465 non-null  object 
 10  price           30758 non-null  object 
 11  early_access    32135 non-null  bool   
 12  id              32133 non-null  object 
 13  developer       28836 non-null  object 
 14  sentiment       24953 non-null  object 
 15  metascore       2677 non-null   object 
dtypes: bool(1), float64(1), object(14)
memory usage: 3.7+ MB


3.c-Consulta de cantidad de nulos por campo

In [553]:
df0.isna().sum() #Consulta de nulos por campo

publisher          8052
genres             3283
app_name              2
title              2050
url                   0
release_date       2067
tags                163
discount_price    31910
reviews_url           2
specs               670
price              1377
early_access          0
id                    2
developer          3299
sentiment          7182
metascore         29458
dtype: int64

**4 - Identificación de features importantes**

4.a - Eliminación de variables consideradas irrelevantes y/o redundantes

In [554]:
columnas_a_eliminar = ['title','url','reviews_url','publisher','developer','id'] #Se va a trabajar con la columna app_name y se elimina la columna title por redundante.
df0.drop(columns=columnas_a_eliminar, inplace=True)  #Las columnas url, reviews_url, publisher, id y developer se eliminan por irrelevantes para el análisis

**5 – Tratamiento de valores faltantes**

5.a - Eliminación de Nan en realease_date

Eliminación de los NaN de la columna release_date. No se identifica manera de imputarlo. La cantidad de registros eliminados es el 6,4% de total de dataset.

In [555]:
df0.dropna(subset = ['release_date'],inplace=True) #Se eliminan los nulos de release_date porque este campo es fundamental y no los podemos imputar de ninguna forma

In [556]:
df0.info()

<class 'pandas.core.frame.DataFrame'>
Index: 30068 entries, 0 to 32133
Data columns (total 10 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   genres          28834 non-null  object 
 1   app_name        30067 non-null  object 
 2   release_date    30068 non-null  object 
 3   tags            29907 non-null  object 
 4   discount_price  204 non-null    float64
 5   specs           29399 non-null  object 
 6   price           28821 non-null  object 
 7   early_access    30068 non-null  bool   
 8   sentiment       23156 non-null  object 
 9   metascore       2623 non-null   object 
dtypes: bool(1), float64(1), object(8)
memory usage: 2.3+ MB


**6 - Creación de la columna release_year necesaria para los endpoints**

In [557]:
df0['release_date'].unique() #Consulta de valores únicos de la columna release_date

array(['2018-01-04', '2017-07-24', '2017-12-07', ..., '2016-11-19',
       'January 2018', '2018-10-01'], dtype=object)

In [558]:
df0['release_year'] = df0['release_date'].str.extract(r'(\d{4})') #Con este método buscamos y extraemos secuencias de cuatro dígitos(Distintos formatos de datos en la columna release_date)

In [559]:
df0.release_year.isna().sum() #Consulta de datos que no se pudieron extraer en el paso previo

101

**7 – Tratamiento de posibles errores**


7.a - Generación de una máscara con los nulos del campo release_year para enviar a un data frame los posibles registros con datos erróneos en el campo release_date

In [560]:
mask = df0['release_year'].isna()    #Se genera una máscara con los nulos del campo release_year
df_filtrado = df0[mask]              #Con la máscara se crea un data frame con los registros con posibles errores en el campo release_date
df_filtrado.head(10)                 #Consulta del data frame - df_filtrado es un data frame auxiliar

genres          app_name  \
10        [Action, Indie, Simulation, Early Access]   Uncanny Islands   
44     [Adventure, Free to Play, RPG, Early Access]    Idle Adventure   
6737                                [Action, Indie]          Warhawks   
7134                 [Adventure, Indie, Simulation]               Eco   
7189                     [Adventure, Casual, Indie]             Woven   
8338                            [Indie, Simulation]        Orcish Inn   
8552                     [Adventure, Casual, Indie]    Out Of The Box   
9335                     [Action, Adventure, Indie]  Attack On Kitten   
10275                 [Free to Play, RPG, Strategy]     Samurai Forge   
10483                                       [Indie]  Edge of Insanity   

         release_date                                               tags  \
10             Soon..  [Early Access, Action, Indie, Simulation, Surv...   
44          Beta测试已开启       [Early Access, Free to Play, Adventure, RPG]   
6737      Coming Soon                        [Action, Indie, VR, Flight]   
7134      Coming Soon  [Survival, Adventure, Multiplayer, Simulation,...   
7189              TBA  [Indie, Casual, Adventure, Singleplayer, Story...   
8338   When it's done                      [Simulation, Indie, Survival]   
8552      coming soon                         [Indie, Adventure, Casual]   
9335              TBA             [Action, Indie, Adventure, Platformer]   
10275            soon                      [Free to Play, Strategy, RPG]   
10483         14 July  [Violent, Indie, Psychological Horror, Horror,...   

       discount_price                                              specs  \
10                NaN                                    [Single-player]   
44                NaN                          [MMO, Steam Achievements]   
6737              NaN  [Single-player, Multi-player, Steam Achievemen...   
7134              NaN  [Single-player, Multi-player, Online Multi-Pla...   
7189              NaN                                    [Single-player]   
8338              NaN                                    [Single-player]   
8552              NaN  [Single-player, Steam Achievements, Steam Trad...   
9335              NaN  [Single-player, Steam Achievements, Full contr...   
10275             NaN  [Single-player, Online Multi-Player, Online Co...   
10483             NaN                                    [Single-player]   

      price  early_access sentiment metascore release_year  
10      NaN          True       NaN       NaN          NaN  
44      NaN          True       NaN       NaN          NaN  
6737    NaN         False       NaN       NaN          NaN  
7134    NaN         False       NaN       NaN          NaN  
7189    NaN         False       NaN       NaN          NaN  
8338    NaN         False       NaN       NaN          NaN  
8552    NaN         False       NaN       NaN          NaN  
9335    NaN         False       NaN       NaN          NaN  
10275   NaN         False       NaN       NaN          NaN  
10483   NaN         False       NaN       NaN          NaN

In [561]:
df_filtrado.tail(10)        #Consulta del data frame

genres  \
25926              [Action, Casual, Indie]   
31925              [Casual, Indie, Racing]   
31943                      [Action, Indie]   
31952                             [Casual]   
31958                             [Action]   
31983                   [Casual, Strategy]   
31989           [Action, Adventure, Indie]   
32000                   [Adventure, Indie]   
32008                      [Action, Indie]   
32087  [Action, Casual, Indie, Simulation]   

                                       app_name     release_date  \
25926                                     Shwip  When it's done!   
31925                                  Sparkour      Coming Soon   
31943                  Acceleration of SUGURI 2      Coming Soon   
31952                              Dungeon Girl      Coming Soon   
31958  Sniper Ghost Warrior 3 - Multiplayer map      coming soon   
31983                            Lost Artifacts      Coming soon   
31989                                   Memento              TBD   
32000                             Spy of Deimos      Coming Soon   
32008                                      DUSK            SOON™   
32087                          Lonely Astronaut      Coming Soon   

                                                    tags  discount_price  \
25926                            [Action, Indie, Casual]             NaN   
31925  [Racing, Parkour, Casual, Indie, Multiplayer, ...             NaN   
31943  [Indie, Action, Bullet Hell, Anime, Cute, Shoo...             NaN   
31952  [Casual, Anime, Visual Novel, Puzzle, Match 3,...             NaN   
31958                                  [Action, Violent]             NaN   
31983                                 [Strategy, Casual]             NaN   
31989                 [Indie, Action, Adventure, Horror]             NaN   
32000                                 [Indie, Adventure]             NaN   
32008  [Action, FPS, Indie, Retro, Shooter, Horror, G...             NaN   
32087                [Casual, Action, Indie, Simulation]             NaN   

                                                   specs  price  early_access  \
25926  [Single-player, Local Multi-Player, Shared/Spl...    NaN         False   
31925  [Single-player, Multi-player, Online Multi-Pla...    NaN         False   
31943  [Single-player, Multi-player, Shared/Split Scr...    NaN         False   
31952  [Single-player, Steam Achievements, Steam Clou...    NaN         False   
31958  [Single-player, Downloadable Content, Partial ...  19.99         False   
31983                [Single-player, Steam Achievements]    NaN         False   
31989                                    [Single-player]    NaN         False   
32000   [Single-player, Steam Achievements, Steam Cloud]    NaN         False   
32008  [Single-player, Online Multi-Player, Steam Ach...   20.0         False   
32087                                    [Single-player]    NaN         False   

      sentiment metascore release_year  
25926       NaN       NaN          NaN  
31925       NaN       NaN          NaN  
31943       NaN       NaN          NaN  
31952       NaN       NaN          NaN  
31958       NaN       NaN          NaN  
31983       NaN       NaN          NaN  
31989       NaN       NaN          NaN  
32000       NaN       NaN          NaN  
32008       NaN       NaN          NaN  
32087       NaN       NaN          NaN

In [562]:
df_filtrado.info()  #Consulta de los campos del data frame filtrado

<class 'pandas.core.frame.DataFrame'>
Index: 101 entries, 10 to 32087
Data columns (total 11 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   genres          101 non-null    object 
 1   app_name        101 non-null    object 
 2   release_date    101 non-null    object 
 3   tags            101 non-null    object 
 4   discount_price  0 non-null      float64
 5   specs           101 non-null    object 
 6   price           3 non-null      object 
 7   early_access    101 non-null    bool   
 8   sentiment       0 non-null      object 
 9   metascore       0 non-null      object 
 10  release_year    0 non-null      object 
dtypes: bool(1), float64(1), object(9)
memory usage: 8.8+ KB


In [563]:
df_filtrado.release_date.nunique()  #Consulta de cantidad de registros únicos en release_date del data frame filtrado

37

In [564]:
df_filtrado.release_date.unique()   #Consulta de lista de registros únicos en release_date del data frame fitlrado

array(['Soon..', 'Beta测试已开启', 'Coming Soon', 'TBA', "When it's done",
       'coming soon', 'soon', '14 July', 'To be Announced', 'TBD', 'Soon',
       'Please wait warmly', 'early access', 'SOON',
       'Coming Soon/Próximamente', 'Coming soon',
       '0̵1̴0̵0̶1̷0̶0̵0̴ ̴0̶0̶1̶1̶0̷0̶1̵1̴ ̸0̶0̶1̶1̵0̶1̷0̴0̵ ̴0̶1̷0̸1̵0̷0̴1̶0̴ ̴0̷0̴1̷1̶0̶1̵1̷1̵ ̵',
       '"""Soon"""', 'Demo coming soon.', 'Not yet available',
       "C'est bientôt...                    (ou pas)",
       'Datachunks conflicted. Be vigiliant.', 'Demo is available now!',
       'Play Beta in demo!', 'Coming Soon!', 'To be announced',
       'When it is finished', 'Release Date TBA', 'When it is ready',
       'Early Access Starting Soon!', 'Comming Soon', 'Early Access soon',
       'To Be Announced', 'Coming soon..', 'Soon™', "When it's done!",
       'SOON™'], dtype=object)

Estos datos se descartan por no tener información que se pueda asociar a un año

In [565]:
del df_filtrado

7.b Limpieza de columna release_date mediante la columna release_year

Se eliminan de esta manera los registros erróneos de release_date

In [566]:
df1 = df0.dropna(subset=['release_year']) #Se eliminan de esta manera los registros erróneos de release_date - df1 data frame del proyecto (Rev.1) sin nulos de release_year

In [567]:
del df0

In [568]:
df1.info() #Consulta del nuevo data frame

<class 'pandas.core.frame.DataFrame'>
Index: 29967 entries, 0 to 32133
Data columns (total 11 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   genres          28733 non-null  object 
 1   app_name        29966 non-null  object 
 2   release_date    29967 non-null  object 
 3   tags            29806 non-null  object 
 4   discount_price  204 non-null    float64
 5   specs           29298 non-null  object 
 6   price           28818 non-null  object 
 7   early_access    29967 non-null  bool   
 8   sentiment       23156 non-null  object 
 9   metascore       2623 non-null   object 
 10  release_year    29967 non-null  object 
dtypes: bool(1), float64(1), object(9)
memory usage: 2.5+ MB


In [569]:
df1.head() #Consulta del nuevo data frame

genres                 app_name  \
0      [Action, Casual, Indie, Simulation, Strategy]      Lost Summoner Kitty   
1               [Free to Play, Indie, RPG, Strategy]                Ironbound   
2  [Casual, Free to Play, Indie, Simulation, Sports]  Real Pool 3D - Poolians   
3                        [Action, Adventure, Casual]                  弹炸人2222   
5                    [Action, Adventure, Simulation]    Battle Royale Trainer   

  release_date                                               tags  \
0   2018-01-04      [Strategy, Action, Indie, Casual, Simulation]   
1   2018-01-04  [Free to Play, Strategy, Indie, RPG, Card Game...   
2   2017-07-24  [Free to Play, Simulation, Sports, Casual, Ind...   
3   2017-12-07                        [Action, Adventure, Casual]   
5   2018-01-04  [Action, Adventure, Simulation, FPS, Shooter, ...   

   discount_price                                              specs  \
0            4.49                                    [Single-player]   
1             NaN  [Single-player, Multi-player, Online Multi-Pla...   
2             NaN  [Single-player, Multi-player, Online Multi-Pla...   
3            0.83                                    [Single-player]   
5             NaN                [Single-player, Steam Achievements]   

          price  early_access        sentiment metascore release_year  
0          4.99         False              NaN       NaN         2018  
1  Free To Play         False  Mostly Positive       NaN         2018  
2  Free to Play         False  Mostly Positive       NaN         2017  
3          0.99         False              NaN       NaN         2017  
5          3.99         False            Mixed       NaN         2018

In [570]:
df1[['release_date','release_year']] #Consulta solo de las columnas release_date y release_year

release_date release_year
0       2018-01-04         2018
1       2018-01-04         2018
2       2017-07-24         2017
3       2017-12-07         2017
5       2018-01-04         2018
...            ...          ...
32129   2018-01-04         2018
32130   2018-01-04         2018
32131   2018-01-04         2018
32132   2018-01-04         2018
32133   2017-09-02         2017

[29967 rows x 2 columns]

**8 - Identificación de features importantes**

8.a - Eliminación de la columna release_date por resultar ya irrelevante para el proyecto

In [571]:
df1.drop(columns='release_date', inplace=True) #Se elimina la columna release_date por resultar ya irrelevante para el proyecto

**9 - Transformaciones**

9.a - Conversión de la variable release_year a numérica

In [572]:
df1['release_year'] = pd.to_numeric(df1['release_year']) #Conversión de la variable release_year a numérica

In [573]:
df1.info()

<class 'pandas.core.frame.DataFrame'>
Index: 29967 entries, 0 to 32133
Data columns (total 10 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   genres          28733 non-null  object 
 1   app_name        29966 non-null  object 
 2   tags            29806 non-null  object 
 3   discount_price  204 non-null    float64
 4   specs           29298 non-null  object 
 5   price           28818 non-null  object 
 6   early_access    29967 non-null  bool   
 7   sentiment       23156 non-null  object 
 8   metascore       2623 non-null   object 
 9   release_year    29967 non-null  int64  
dtypes: bool(1), float64(1), int64(1), object(7)
memory usage: 2.3+ MB


9.b - Expansión de las columnas anidadas

Se desanida el campo genres

In [574]:
df2 = df1.explode('genres') #Se desanida el campo genres - df2 es data frame de proyecto (Rev.2) con la columna genres desanidada


In [575]:
del df1

In [576]:
df2.head(10) #Consulta del nuevo data frame desanidado

genres                 app_name  \
0        Action      Lost Summoner Kitty   
0        Casual      Lost Summoner Kitty   
0         Indie      Lost Summoner Kitty   
0    Simulation      Lost Summoner Kitty   
0      Strategy      Lost Summoner Kitty   
1  Free to Play                Ironbound   
1         Indie                Ironbound   
1           RPG                Ironbound   
1      Strategy                Ironbound   
2        Casual  Real Pool 3D - Poolians   

                                                tags  discount_price  \
0      [Strategy, Action, Indie, Casual, Simulation]            4.49   
0      [Strategy, Action, Indie, Casual, Simulation]            4.49   
0      [Strategy, Action, Indie, Casual, Simulation]            4.49   
0      [Strategy, Action, Indie, Casual, Simulation]            4.49   
0      [Strategy, Action, Indie, Casual, Simulation]            4.49   
1  [Free to Play, Strategy, Indie, RPG, Card Game...             NaN   
1  [Free to Play, Strategy, Indie, RPG, Card Game...             NaN   
1  [Free to Play, Strategy, Indie, RPG, Card Game...             NaN   
1  [Free to Play, Strategy, Indie, RPG, Card Game...             NaN   
2  [Free to Play, Simulation, Sports, Casual, Ind...             NaN   

                                               specs         price  \
0                                    [Single-player]          4.99   
0                                    [Single-player]          4.99   
0                                    [Single-player]          4.99   
0                                    [Single-player]          4.99   
0                                    [Single-player]          4.99   
1  [Single-player, Multi-player, Online Multi-Pla...  Free To Play   
1  [Single-player, Multi-player, Online Multi-Pla...  Free To Play   
1  [Single-player, Multi-player, Online Multi-Pla...  Free To Play   
1  [Single-player, Multi-player, Online Multi-Pla...  Free To Play   
2  [Single-player, Multi-player, Online Multi-Pla...  Free to Play   

   early_access        sentiment metascore  release_year  
0         False              NaN       NaN          2018  
0         False              NaN       NaN          2018  
0         False              NaN       NaN          2018  
0         False              NaN       NaN          2018  
0         False              NaN       NaN          2018  
1         False  Mostly Positive       NaN          2018  
1         False  Mostly Positive       NaN          2018  
1         False  Mostly Positive       NaN          2018  
1         False  Mostly Positive       NaN          2018  
2         False  Mostly Positive       NaN          2017

In [577]:
df2.info() #Consulta del nuevo data frame desanidado

<class 'pandas.core.frame.DataFrame'>
Index: 72450 entries, 0 to 32133
Data columns (total 10 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   genres          71216 non-null  object 
 1   app_name        72448 non-null  object 
 2   tags            72267 non-null  object 
 3   discount_price  580 non-null    float64
 4   specs           71510 non-null  object 
 5   price           69542 non-null  object 
 6   early_access    72450 non-null  bool   
 7   sentiment       56152 non-null  object 
 8   metascore       5474 non-null   object 
 9   release_year    72450 non-null  int64  
dtypes: bool(1), float64(1), int64(1), object(7)
memory usage: 5.6+ MB


In [578]:
lista_genres = df2.genres.unique() #Consulta de valores únicos del campo genres
lista_genres

array(['Action', 'Casual', 'Indie', 'Simulation', 'Strategy',
       'Free to Play', 'RPG', 'Sports', 'Adventure', 'Racing', nan,
       'Massively Multiplayer', 'Early Access',
       'Animation &amp; Modeling', 'Video Production', 'Web Publishing',
       'Education', 'Software Training', 'Utilities',
       'Design &amp; Illustration', 'Audio Production', 'Photo Editing',
       'Accounting'], dtype=object)

In [579]:
df2.genres.nunique() #Consulta de la cantidad de valores únicos del campo genres

22

9.c – Corrección de registros

Corrección de los valores 'Animation &amp; Modeling' y 'Design &amp; Illustration' en los registros

In [580]:
df2['genres'] = df2['genres'].replace({'Animation &amp; Modeling':'Animation & Modeling',
                                                            'Design &amp; Illustration':'Design & Illustration'}) #Se corrigen dos valores en los registros

In [581]:
df2.isna().sum() #Consulta de nulos del data frame

genres             1234
app_name              2
tags                183
discount_price    71870
specs               940
price              2908
early_access          0
sentiment         16298
metascore         66976
release_year          0
dtype: int64

**10 – Imputaciones**

10.a - Imputación de valores nulos

Se imputan de valores nulos del campo genres con valores coincidentes de genres en el campo tags

Se crea la columna imputación para registrar que registros fueron imputados en el campo genres - 0 para valor no imputado y 1 para valor imputado

In [582]:
df2['imputacion'] = df2['genres'].isna().astype(int) #Creacion de la columna imputación - 1 para valores Nan

In [583]:
df2.info() #Consulta del data frame

<class 'pandas.core.frame.DataFrame'>
Index: 72450 entries, 0 to 32133
Data columns (total 11 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   genres          71216 non-null  object 
 1   app_name        72448 non-null  object 
 2   tags            72267 non-null  object 
 3   discount_price  580 non-null    float64
 4   specs           71510 non-null  object 
 5   price           69542 non-null  object 
 6   early_access    72450 non-null  bool   
 7   sentiment       56152 non-null  object 
 8   metascore       5474 non-null   object 
 9   release_year    72450 non-null  int64  
 10  imputacion      72450 non-null  int32  
dtypes: bool(1), float64(1), int32(1), int64(1), object(7)
memory usage: 5.9+ MB


In [584]:
df2.head() #Consulta del data frame

genres             app_name  \
0      Action  Lost Summoner Kitty   
0      Casual  Lost Summoner Kitty   
0       Indie  Lost Summoner Kitty   
0  Simulation  Lost Summoner Kitty   
0    Strategy  Lost Summoner Kitty   

                                            tags  discount_price  \
0  [Strategy, Action, Indie, Casual, Simulation]            4.49   
0  [Strategy, Action, Indie, Casual, Simulation]            4.49   
0  [Strategy, Action, Indie, Casual, Simulation]            4.49   
0  [Strategy, Action, Indie, Casual, Simulation]            4.49   
0  [Strategy, Action, Indie, Casual, Simulation]            4.49   

             specs price  early_access sentiment metascore  release_year  \
0  [Single-player]  4.99         False       NaN       NaN          2018   
0  [Single-player]  4.99         False       NaN       NaN          2018   
0  [Single-player]  4.99         False       NaN       NaN          2018   
0  [Single-player]  4.99         False       NaN       NaN          2018   
0  [Single-player]  4.99         False       NaN       NaN          2018   

   imputacion  
0           0  
0           0  
0           0  
0           0  
0           0

10.b - Creación de un nuevo data frame con una máscara para guardar los registros nulos de genres

In [585]:
mask = df2['genres'].isna()     #Se crea una máscara para valores nulos de genres
df_genres_nan = df2[mask]       #Se crea un nuevo data frame con los registros nulos de genres - df_genres_nan es un data frame auxiliar

10.c - Eliminación de nulos en el data frame principal

Se eliminan los nulos del campo genres

In [586]:
df3 = df2.dropna(subset=['genres']) #Se eliminan los nulos del campo genres - df3 es el data frame de proyecto (Rev. 3) con nulos de genres eliminados

In [587]:
del df2

In [588]:
df_genres_nan.info()    #Consulta del data frame de nulos de genres

<class 'pandas.core.frame.DataFrame'>
Index: 1234 entries, 26 to 32121
Data columns (total 11 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   genres          0 non-null      object 
 1   app_name        1234 non-null   object 
 2   tags            1097 non-null   object 
 3   discount_price  0 non-null      float64
 4   specs           851 non-null    object 
 5   price           1203 non-null   object 
 6   early_access    1234 non-null   bool   
 7   sentiment       624 non-null    object 
 8   metascore       13 non-null     object 
 9   release_year    1234 non-null   int64  
 10  imputacion      1234 non-null   int32  
dtypes: bool(1), float64(1), int32(1), int64(1), object(7)
memory usage: 102.4+ KB


10.d - Desanidado del campo tags en el data frame auxiliar df_genres_nan y re indexado

In [589]:
df_genres_nan = df_genres_nan.explode('tags')   #Desanidado del campo tags en el data frame auxiliar df_genres_nan

In [590]:
df_genres_nan.reset_index(drop=True, inplace=True)   #Se reindexa el data frame auxiliar

10.e - Proceso de imputación de datos en genres

In [591]:
registros = df_genres_nan.shape   #Guardado de la tupla con la información del data frame auxiliar

In [592]:
lista_genres = df3.genres.unique()    #Generación de una lista de géneros del campo genres para comparación
lista_genres

array(['Action', 'Casual', 'Indie', 'Simulation', 'Strategy',
       'Free to Play', 'RPG', 'Sports', 'Adventure', 'Racing',
       'Massively Multiplayer', 'Early Access', 'Animation & Modeling',
       'Video Production', 'Web Publishing', 'Education',
       'Software Training', 'Utilities', 'Design & Illustration',
       'Audio Production', 'Photo Editing', 'Accounting'], dtype=object)

10.f - Definición de una función para obtención de los datos de imputación

In [593]:
def imputacion(tag,lista):              #En esta función ingresa un dato y se verifica si está contenido en una lista.
    genero_imputado = 'Sin Datos'       #En caso afirmativo, devuelve el valor de tag, contrario, devuelve Sin Datos
    if tag in lista:
        genero_imputado = tag
    return genero_imputado     

10.g – Imputación de valores en genres

In [594]:
for i in range(0,registros[0]):                 #Con este ciclo for se imputan valores en genres con los datos de tags en el data frame auxiliar df_genres_nan
    tag = df_genres_nan.loc[i,'tags']
    genero = imputacion(tag,lista_genres)
    df_genres_nan.loc[i, 'genres'] = genero

In [595]:
df_genres_nan.head()

genres                                           app_name  \
0  Design & Illustration  Robotpencil Presents: Understanding 3D for Con...   
1              Sin Datos  Robotpencil Presents: Understanding 3D for Con...   
2                 Casual                                  Dr. Daisy Pet Vet   
3                 Casual                                           Hot Dish   
4              Sin Datos        Call of Cthulhu®: Dark Corners of the Earth   

                    tags  discount_price            specs price  early_access  \
0  Design & Illustration             NaN              NaN  4.99         False   
1               Tutorial             NaN              NaN  4.99         False   
2                 Casual             NaN  [Single-player]  9.99         False   
3                 Casual             NaN  [Single-player]  9.99         False   
4                 Horror             NaN  [Single-player]  4.99         False   

         sentiment metascore  release_year  imputacion  
0              NaN       NaN          2018           1  
1              NaN       NaN          2018           1  
2   3 user reviews        NA          2008           1  
3   4 user reviews        NA          2008           1  
4  Mostly Positive        76          2006           1

10.h – Eliminación del campo tags

Se elimina el campo tags en el data frame del proyecto df3 y en el auxiliar df_genres_nan por irrelevante

In [596]:
df_genres_nan.drop('tags', axis=1, inplace=True) 

In [597]:
df3.drop('tags', axis=1, inplace=True)

10.i - Concatenación del data frame principal del proyecto y el data frame auxiliar

In [598]:
df4 = pd.concat([df3, df_genres_nan]) #Concatenación de df3 y df_genres_nan en df4 - df4 data frame de proyecto (Rev. 4)

In [599]:
del df3
del df_genres_nan

**11 - Transformaciones**

11.a - Desanidado de campo specs

In [600]:
df5 = df4.explode('specs')  #Desanidado del campo specs en un nuevo data frame - df5 data frame de proyecto (Rev. 5) con el campo specs desanidado

In [601]:
del df4

In [602]:
df5.reset_index(drop=True, inplace=True)    #Reindexado del data frame de proyecto

In [603]:
df5.info() #Consulta del data frame de proyecto

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 324469 entries, 0 to 324468
Data columns (total 10 columns):
 #   Column          Non-Null Count   Dtype  
---  ------          --------------   -----  
 0   genres          324469 non-null  object 
 1   app_name        324465 non-null  object 
 2   discount_price  2701 non-null    float64
 3   specs           322824 non-null  object 
 4   price           312696 non-null  object 
 5   early_access    324469 non-null  bool   
 6   sentiment       247241 non-null  object 
 7   metascore       25850 non-null   object 
 8   release_year    324469 non-null  int64  
 9   imputacion      324469 non-null  int32  
dtypes: bool(1), float64(1), int32(1), int64(1), object(6)
memory usage: 21.4+ MB


11.b - Consulta de datos únicos del campo specs

In [604]:
lista_specs = df5.specs.unique()    #Consulta de datos únicos del campo specs
lista_specs

array(['Single-player', 'Multi-player', 'Online Multi-Player',
       'Cross-Platform Multiplayer', 'Steam Achievements',
       'Steam Trading Cards', 'In-App Purchases', 'Stats',
       'Downloadable Content', 'Full controller support', 'Steam Cloud',
       'Steam Leaderboards', 'Partial Controller Support', 'Local Co-op',
       'Shared/Split Screen', 'Valve Anti-Cheat enabled',
       'Local Multi-Player', 'Steam Turn Notifications', 'MMO', 'Co-op',
       'Captions available', 'Steam Workshop', 'Includes level editor',
       'Mods', 'Mods (require HL2)', 'Online Co-op', 'Game demo',
       'Includes Source SDK', 'Commentary available', nan,
       'SteamVR Collectibles', 'Mods (require HL1)'], dtype=object)

In [605]:
df5.specs.nunique() #Consulta de cantidad de datos únicos del campo specs

31

In [606]:
df5.head()  #Consulta del data frame de proyecto

genres             app_name  discount_price          specs price  \
0      Action  Lost Summoner Kitty            4.49  Single-player  4.99   
1      Casual  Lost Summoner Kitty            4.49  Single-player  4.99   
2       Indie  Lost Summoner Kitty            4.49  Single-player  4.99   
3  Simulation  Lost Summoner Kitty            4.49  Single-player  4.99   
4    Strategy  Lost Summoner Kitty            4.49  Single-player  4.99   

   early_access sentiment metascore  release_year  imputacion  
0         False       NaN       NaN          2018           0  
1         False       NaN       NaN          2018           0  
2         False       NaN       NaN          2018           0  
3         False       NaN       NaN          2018           0  
4         False       NaN       NaN          2018           0

11.c - Reemplazo de nulos del campo specs por Sin Datos

In [607]:
df5.specs.fillna('Sin Datos', inplace=True) #Reemplazo de nulos del campo specs por Sin Datos

11.d - Conversión del campo price a numérico y reemplazo de errores por Nan

In [608]:
df5.isna().sum()    #Consulta de nulos del data frame de proyecto

genres                 0
app_name               4
discount_price    321768
specs                  0
price              11773
early_access           0
sentiment          77228
metascore         298619
release_year           0
imputacion             0
dtype: int64

In [609]:
df5['price'].unique()

array([4.99, 'Free To Play', 'Free to Play', 0.99, 3.99, 9.99, 18.99,
       29.99, nan, 10.99, 2.99, 1.59, 14.99, 1.99, 59.99, 8.99, 6.99,
       7.99, 39.99, 'Free', 19.99, 7.49, 12.99, 5.99, 2.49, 15.99, 1.25,
       24.99, 17.99, 61.99, 3.49, 11.99, 13.99, 'Free Demo',
       'Play for Free!', 34.99, 1.49, 32.99, 99.99, 14.95, 69.99, 16.99,
       79.99, 49.99, 5.0, 44.99, 13.98, 29.96, 109.99, 149.99, 771.71,
       'Install Now', 21.99, 89.99, 'Play WARMACHINE: Tactics Demo', 0.98,
       139.92, 4.29, 64.99, 'Free Mod', 54.99, 74.99, 'Install Theme',
       0.89, 'Third-party', 0.5, 'Play Now', 299.99, 1.29, 119.99, 3.0,
       15.0, 5.49, 23.99, 49.0, 20.99, 10.93, 1.39,
       'Free HITMAN™ Holiday Pack', 36.99, 4.49, 2.0, 4.0, 1.95, 1.5,
       199.0, 189.0, 6.66, 27.99, 129.99, 179.0, 26.99, 399.99, 31.99,
       399.0, 20.0, 40.0, 3.33, 22.99, 320.0, 38.85, 71.7, 995.0, 27.49,
       3.39, 6.0, 19.95, 499.99, 199.99, 16.06, 4.68, 131.4, 44.98,
       202.76, 2.3, 0.95, 172.

In [610]:
df5['price'] = pd.to_numeric(df5['price'], errors='coerce') #Conversión del campo price a numérico y reemplazo de errores por Nan

In [611]:
df5.isna().sum() #Consulta de nulos del data frame de proyecto

genres                 0
app_name               4
discount_price    321768
specs                  0
price              32457
early_access           0
sentiment          77228
metascore         298619
release_year           0
imputacion             0
dtype: int64

**12 - Revisiones**

12.a - Revisión de campo early_access

In [612]:
df5.early_access.nunique()  #Revisión de cantidad de valores únicos del campo early_access (Valores False y True)

2

12.b - Revisión del campo sentiment

In [613]:
df5.sentiment.nunique()   #Revisión de cantidad de valores únicos del campo sentimentdf5.sentiment.nunique() 

18

In [614]:
lista_sentiment = df5.sentiment.unique() #Revisión de datos unicos del campo sentiment
lista_sentiment

array([nan, 'Mostly Positive', 'Mixed', '1 user reviews',
       '3 user reviews', 'Very Positive', 'Overwhelmingly Positive',
       '6 user reviews', '5 user reviews', '2 user reviews',
       'Very Negative', 'Positive', 'Mostly Negative', '9 user reviews',
       '8 user reviews', 'Negative', '7 user reviews', '4 user reviews',
       'Overwhelmingly Negative'], dtype=object)

Reemplazo de datos en el campo sentiment - Los valores Nan y que contengan "user reviews" se reemplazan por "Sin Datos" porque no aplican como calificación

In [615]:
df5['sentiment'] = df5['sentiment'].replace({None: 'Sin Datos', 
                                    '1 user reviews':'Sin Datos',
                                    '3 user reviews':'Sin Datos',
                                    '6 user reviews':'Sin Datos',
                                    '5 user reviews':'Sin Datos',
                                    '2 user reviews':'Sin Datos',
                                    '9 user reviews':'Sin Datos',
                                    '8 user reviews':'Sin Datos',
                                    '7 user reviews':'Sin Datos',
                                    '4 user reviews':'Sin Datos'})  #Reemplazo de valores Nan y que contengan "user reviews" se reemplazan por "Sin Datos"

12.c - Revisión del campo metascore

In [616]:
df5.metascore.nunique() #Revisión de cantidad de valores únicos del campo metascore

71

In [617]:
lista_metascore = df5.metascore.unique()  #Revisión de datos únicos del campo metascore
lista_metascore

array([nan, 96, 84, 80, 76, 70, 'NA', 69, 81, 75, 72, 66, 67, 77, 91, 89,
       83, 61, 88, 65, 94, 57, 86, 87, 92, 79, 82, 58, 74, 85, 90, 68, 71,
       60, 73, 59, 64, 54, 53, 78, 51, 44, 63, 38, 56, 49, 52, 62, 93, 48,
       34, 95, 43, 55, 24, 46, 41, 20, 39, 45, 35, 47, 40, 36, 50, 32, 37,
       33, 42, 27, 29, 30], dtype=object)

Conversión del campo metascore a numérico y de posibles errores a Nan (p.e. valor "NA")

In [618]:
df5['metascore'] = pd.to_numeric(df5['metascore'], errors='coerce') #Conversión del campo metascore a numérico y de posibles errores a Nan (p.e. valor "NA")

Reindexado del data frame de proyecto

In [619]:
df5.reset_index(drop=True, inplace=True) #Reindexado del data frame de proyecto

In [620]:
df5.head() #Consulta del data frama de proyecto

genres             app_name  discount_price          specs  price  \
0      Action  Lost Summoner Kitty            4.49  Single-player   4.99   
1      Casual  Lost Summoner Kitty            4.49  Single-player   4.99   
2       Indie  Lost Summoner Kitty            4.49  Single-player   4.99   
3  Simulation  Lost Summoner Kitty            4.49  Single-player   4.99   
4    Strategy  Lost Summoner Kitty            4.49  Single-player   4.99   

   early_access  sentiment  metascore  release_year  imputacion  
0         False  Sin Datos        NaN          2018           0  
1         False  Sin Datos        NaN          2018           0  
2         False  Sin Datos        NaN          2018           0  
3         False  Sin Datos        NaN          2018           0  
4         False  Sin Datos        NaN          2018           0

In [621]:
df5.info()  #Consulta del data frama de proyecto

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 324469 entries, 0 to 324468
Data columns (total 10 columns):
 #   Column          Non-Null Count   Dtype  
---  ------          --------------   -----  
 0   genres          324469 non-null  object 
 1   app_name        324465 non-null  object 
 2   discount_price  2701 non-null    float64
 3   specs           324469 non-null  object 
 4   price           292012 non-null  float64
 5   early_access    324469 non-null  bool   
 6   sentiment       324469 non-null  object 
 7   metascore       25616 non-null   float64
 8   release_year    324469 non-null  int64  
 9   imputacion      324469 non-null  int32  
dtypes: bool(1), float64(3), int32(1), int64(1), object(4)
memory usage: 21.4+ MB


In [622]:
df5.isna().sum()

genres                 0
app_name               4
discount_price    321768
specs                  0
price              32457
early_access           0
sentiment              0
metascore         298853
release_year           0
imputacion             0
dtype: int64

12.d - Revisión del campo app_name

Se reemplazan nulos en el campo app_name

In [623]:
df5['app_name'] = df5['app_name'].fillna('Sin Datos')   #Reemplazo de nulos por "Sin Datos"

In [624]:
df5.isna().sum()    #Consulta de nulos del data frame de proyecto

genres                 0
app_name               0
discount_price    321768
specs                  0
price              32457
early_access           0
sentiment              0
metascore         298853
release_year           0
imputacion             0
dtype: int64

**13 - Exportación**

Se almacena en formato parquet por su almacenamiento columnar y ventajas en el uso de funciones de agregación que serán requeridas en los enpoints

In [625]:
parquet_filename = 'steam_games.parquet'        #Nombre del archivo Parquet

table = pa.Table.from_pandas(df5)               #Convertir el DataFrame a una tabla Arrow

pq.write_table(table, parquet_filename)         #Guardar la tabla Arrow en formato Parquet

**14 - Endpoints**

14.a - def genero( Año: str ): 

Se ingresa un año y devuelve una lista con los 5 géneros más vendidos en el orden correspondiente.

In [626]:
def genero(Anio:str): 
    #Se ingresa un año y devuelve una lista con los 5 géneros más vendidos en el orden correspondiente.
    Anio = int(Anio)
    lista_generos = []
    top_generos = []
    try:
        df_filtrado_anio = df5[df5['release_year'] == Anio]     #Tratamiento de la excepción
    except (ValueError, KeyError, TypeError):
        return None

    df_filtrado_anio_unique = df_filtrado_anio.drop_duplicates(subset=['genres','app_name'])

    # Agrupar por la columna 'Nombre'
    grupos_genres = df_filtrado_anio_unique.groupby('genres')

    # Calcular la media de edad para cada grupo
    top_genres = grupos_genres['genres'].count()
    
    top_genres.sort_values(ascending=False, inplace=True)
    
    # Tomar los primeros 5 elementos de la Serie y almacenarlos en una lista
    lista_generos = top_genres.index.tolist()

    top_generos = lista_generos[:5]

    return {'Anio': Anio, 'Generos': top_generos}

In [627]:
genero(2010)

{'Anio': 2010,
 'Generos': ['Action', 'Indie', 'Strategy', 'Adventure', 'Casual']}

14.b - def juegos( Año: str ): 

Se ingresa un año y devuelve una lista con los juegos lanzados en el año.

In [628]:
def juegos(Anio:str): 
    #Se ingresa un año y devuelve una lista con los juegos lanzados en el año.
    Anio = int(Anio)
    lista_juegos = []
    try:
        df_filtrado_anio = df5[df5['release_year'] == Anio]     #Tratamiento de la excepción
    except (ValueError, KeyError, TypeError):
        return None

    df_filtrado_anio_unique = df_filtrado_anio.drop_duplicates(subset='app_name')

    lista_juegos = list(df_filtrado_anio_unique['app_name'])

    return {'Anio': Anio, 'Juegos': lista_juegos}

In [629]:
juegos(2011)

{'Anio': 2011,
 'Juegos': ['Mafia II',
  'Lara Croft GoL: Raziel and Kain Character Pack',
  'Lara Croft GoL: Things that Go Boom - Challenge Pack 2',
  'Lara Croft GoL: Hazardous Reunion - Challenge Pack 3',
  'Lara Croft GoL: All the Trappings - Challenge Pack 1',
  'Zeit²',
  'A.R.E.S.: Extinction Agenda',
  'APOX',
  'Dead Space™ 2',
  'Drawn®: The Painted Tower',
  'Kaptain Brawe: A Brawe New World',
  'Monday Night Combat',
  'Luxor: 5th Passage',
  'Alien Soldier',
  'Shining in the Darkness',
  'Light Crusader',
  'Landstalker: The Treasures of King Nole',
  'Gunstar Heroes',
  'ToeJam & Earl',
  'Streets of Rage 2',
  'Sonic The Hedgehog 2',
  'Streets of Rage',
  'Shining Force II',
  'Shining Force',
  'Ship Simulator Extremes: Sigita Pack',
  'Trino',
  "Mishap 2: An Intentional Haunting - Collector's Edition",
  'The Sims™ 3 Outdoor Living Stuff',
  "Who's That Flying?!",
  'Myst: Masterpiece Edition',
  'Guardians of Graxia: Elves & Dwarves',
  'Two Worlds II HD',
  'YOU 

14.c - def specs( Año: str ): 

Se ingresa un año y devuelve una lista con los 5 specs que más se repiten en el mismo en el orden correspondiente.

In [630]:
def specs(Anio:str): 
    #Se ingresa un año y devuelve una lista con los 5 géneros más vendidos en el orden correspondiente.
    Anio = int(Anio)
    lista_specs = []
    top_specs = []
    try:
        df_filtrado_anio = df5[df5['release_year'] == Anio]     #Tratamiento de la excepción
    except (ValueError, KeyError, TypeError):
        return None

    df_filtrado_anio_unique = df_filtrado_anio.drop_duplicates(subset=['specs','app_name'])

    # Agrupar por la columna 'Nombre'
    grupos_specs = df_filtrado_anio_unique.groupby('specs')

    # Calcular la media de edad para cada grupo
    top_specs = grupos_specs['specs'].count()

    top_specs.sort_values(ascending=False, inplace=True)

    # Tomar los primeros 5 elementos de la Serie y almacenarlos en una lista
    lista_specs = top_specs.index.tolist()

    top_specs = lista_specs[:5]

    return {'Anio': Anio, 'Specs': top_specs}

In [631]:
specs(2018)

{'Anio': 2018,
 'Specs': ['Single-player',
  'Steam Achievements',
  'Full controller support',
  'Steam Cloud',
  'Partial Controller Support']}

14.d - def earlyacces( Año: str ): 

Cantidad de juegos lanzados en un año con early access.

In [632]:
def earlyacces(Anio:str):        
        #Ingresa un año y devuelve la cantidad de juegos lanzados en un año con early access.
    Anio = int(Anio)
    lista_juegos = []
    try:
        df_filtrado_anio = df5[df5['release_year'] == Anio]     #Tratamiento de la excepción
    except (ValueError, KeyError, TypeError):
        return None

    df_filtrado_anio_unique = df_filtrado_anio.drop_duplicates(subset='app_name')

    condicion = df_filtrado_anio_unique['early_access'] == True

    lista_juegos = list(df_filtrado_anio_unique.loc[condicion,'app_name'])

    cantidad = len(lista_juegos)

    return {'Anio': Anio, 'Cantidad de Juegos con early access': cantidad}

In [633]:
earlyacces(2013)

{'Anio': 2013, 'Cantidad de Juegos con early access': 10}

14.c - def sentiment( Año: str ): 

Según el año de lanzamiento, se devuelve una lista con la cantidad de registros que se encuentren categorizados con un análisis de sentimiento.

                    Ejemplo de retorno: {Mixed = 182, Very Positive = 120, Positive = 278}


In [634]:
def sentiment(Anio:str): 
    #Según el año de lanzamiento, se devuelve una lista con la cantidad de registros que se encuentren categorizados con un análisis de sentimiento.
    Anio = int(Anio)
    
    try:
        df_filtrado_anio = df5[df5['release_year'] == Anio]     #Tratamiento de la excepción
    except (ValueError, KeyError, TypeError):
        return None

    df_filtrado_anio_unique = df_filtrado_anio.drop_duplicates(subset='app_name')

    # Contar la cantidad de veces que se repite cada nombre en la columna 'nombre'
    conteo = df_filtrado_anio_unique['sentiment'].value_counts()

    diccionario = conteo.to_dict()

    return {'Anio': Anio, 'Cantidad de registros categorizados': diccionario}

In [635]:
sentiment(2018)

{'Anio': 2018,
 'Cantidad de registros categorizados': {'Sin Datos': 146,
  'Mixed': 6,
  'Mostly Positive': 3,
  'Very Positive': 3,
  'Mostly Negative': 1}}

14.d - def metascore( Año: str ): 

Top 5 juegos según año con mayor metascore.

In [636]:
def metascore(Anio:str): 
    #Top 5 juegos según año con mayor metascore.
    Anio = int(Anio)
    
    try:
        df_filtrado_anio = df5[df5['release_year'] == Anio]     #Tratamiento de la excepción
    except (ValueError, KeyError, TypeError):
        return 

    df_filtrado_anio_unique = df_filtrado_anio.drop_duplicates(subset='app_name')

    df_filtrado_anio_unique.dropna(subset=['metascore'], inplace=True)

    df_ordenado = df_filtrado_anio_unique.sort_values(by='metascore', ascending=False)

    lista_app_name = df_ordenado['app_name'].to_list()

    top_app_name = lista_app_name[:5]

    return {'Anio': Anio, 'Top 5 de juegos con mayor metascore': top_app_name}

In [637]:
metascore(2015)

C:\Users\guill\AppData\Local\Temp\ipykernel_30076\3249005981.py:12: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_filtrado_anio_unique.dropna(subset=['metascore'], inplace=True)


{'Anio': 2015,
 'Top 5 de juegos con mayor metascore': ['Grand Theft Auto V',
  'Divinity: Original Sin - Enhanced Edition',
  'Undertale',
  'METAL GEAR SOLID V: THE PHANTOM PAIN',
  'Pillars of Eternity']}